db/migration/schema_set.go

```diff
@@ -0,0 +1,219 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migration
+
+import (
+	"database/sql"
+	"errors"
+)
+
+type stringSet map[string]bool
+
+type schemaSetBackend interface {
+	loadSchemaIDs() (stringSet, error)
+	persistSchemaID(schemaID string, tx *sql.Tx) error
+	deleteSchemaID(schemaID string) error
+}
+
+var errNilBackend = errors.New("schemaSetBackend is nil")
+
+type schemaSet struct {
+	schemas stringSet
+	backend schemaSetBackend
+}
+
+func newSchemaSet() *schemaSet {
+	return &schemaSet{
+		schemas: stringSet{},
+		backend: nil, // there is a global schemaSet instance, backend must be lazily initialized
+	}
+}
+
+func (s *schemaSet) init(backend schemaSetBackend) (err error) {
+	log.Notice("post-migration: initializing schema ID set")
+
+	if backend == nil {
+		return errNilBackend
+	}
+
+	s.backend = backend
+
+	schemas, err := backend.loadSchemaIDs()
+	if err != nil {
+		return err
+	}
+
+	s.schemas = schemas
+
+	return nil
+}
+
+func (s *schemaSet) markSchemaID(schemaID string, tx *sql.Tx) (err error) {
+	log.Info("post-migration: marking schema '%s'", schemaID)
+
+	if s.backend == nil {
+		return errNilBackend
+	}
+
+	err = s.backend.persistSchemaID(schemaID, tx)
+	if err != nil {
+		return err
+	}
+
+	s.schemas[schemaID] = true
+	return nil
+}
+
+func (s *schemaSet) getSchemaIDs() []string {
+	schemas := make([]string, 0, len(s.schemas))
+	for schema := range s.schemas {
+		schemas = append(schemas, schema)
+	}
+	return schemas
+}
+
+func (s *schemaSet) removeSchemaID(schemaID string) error {
+	log.Info("post-migration: removing schema '%s'", schemaID)
+
+	if s.backend == nil {
+		return errNilBackend
+	}
+
+	err := s.backend.deleteSchemaID(schemaID)
+	if err != nil {
+		return err
+	}
+
+	delete(s.schemas, schemaID)
+	return nil
+}
+
+type postMigrationEventsBackend struct {
+	db *sql.DB
+}
+
+var _ schemaSetBackend = &postMigrationEventsBackend{}
+
+func newSchemaSetDbBackend(db *sql.DB) *postMigrationEventsBackend {
+	return &postMigrationEventsBackend{db: db}
+}
+
+const schemaSetDbTable = "post_migration_events"
+
+func (b *postMigrationEventsBackend) loadSchemaIDs() (stringSet, error) {
+	log.Debug("post-migration: preloading pending schema IDs")
+
+	tx, err := b.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	schemas, err := b.readSchemaIDsInTx(tx)
+	if err != nil {
+		tx.Rollback()
+		return nil, err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return nil, err
+	}
+
+	return schemas, nil
+}
+
+func (b *postMigrationEventsBackend) readSchemaIDsInTx(tx *sql.Tx) (stringSet, error) {
+	if err := b.ensureTableExists(tx); err != nil {
```

Comment:
I don't like that fact, that function called `read` can modify db state

Reply:
I can extract it, but that will complicate uses (only 2 so not really important) and tests. Your call whether I should do this.

Reply:
I see only 1 use

Reply:
I was thinking about something else, nevermind. Done.